rush-plugins/rush-operation-resource-plugin/src/RushOperationResourcePlugin.ts

```diff
@@ -0,0 +1,214 @@
+import type {
+  IRushPlugin,
+  IPhasedCommand,
+  RushSession,
+  RushConfiguration,
+  ILogger,
+  IOperationRunnerContext,
+  IOperationExecutionResult,
+  OperationStatus,
+  Operation
+} from '@rushstack/rush-sdk';
+import { Stopwatch } from '@rushstack/rush-sdk/lib/utilities/Stopwatch';
+import * as fs from 'fs';
+
+export interface IRushOperationResourcePluginOptions {
+  resourceConstraints?: IResourceConstraintConfig[];
+  resourcePools?: IResourcePoolConfig[];
+}
+
+export interface IResourceConstraintConfig {
+  appliesTo: {
+    operationName?: string;
+    phaseName?: string;
+    packageName?: string;
+    tagName?: string;
+  };
+  resourcePool: {
+    poolName: string;
+    envVarName?: string;
+  };
+}
+
+export interface IResourcePoolConfig {
+  poolName: string;
+  resources?: string[];
+  resourceCount?: number;
+}
+
+export interface IResourcePool {
+  poolName: string;
+  resources: string[];
+}
+
+export const RESOURCE_POOL_POLL_INTERVAL_MS = 500;
+export const PLUGIN_NAME: string = 'RushOperationResourcePlugin';
+
+export class RushOperationResourcePlugin implements IRushPlugin {
+  public readonly pluginName: string = PLUGIN_NAME;
+  private readonly _options: IRushOperationResourcePluginOptions;
+  private readonly _checkedOutResources: Map<Operation, string>;
+  private readonly _resourcePools: Map<string, IResourcePool>;
+
+  public constructor(options: IRushOperationResourcePluginOptions) {
+    this._options = options;
+    this._checkedOutResources = new Map();
+    this._resourcePools = new Map();
+  }
+
+  public apply(rushSession: RushSession, rushContext: RushConfiguration): void {
+    const logger: ILogger = rushSession.getLogger(this.pluginName);
+
+    this._validateOptions();
+
+    rushSession.hooks.runAnyPhasedCommand.tapPromise(
+      this.pluginName,
+      async (command: IPhasedCommand): Promise<void> => {
+        command.hooks.beforeExecuteOperation.tapPromise(
+          this.pluginName,
+          async (
+            runnerContext: IOperationRunnerContext & IOperationExecutionResult
+          ): Promise<OperationStatus | undefined> => {
+            const constraint: IResourceConstraintConfig | undefined = this._getConstraintsForOperation(
+              runnerContext.operation
+            );
+
+            if (constraint) {
+              if (constraint.resourcePool) {
+                const pool: IResourcePool = this._resourcePools.get(constraint.resourcePool.poolName)!;
+
+                // Wait to obtain a resource from the resource pool for this constraint
+                const resource: string = await this._obtainResource(pool, runnerContext.operation);
```

Comment:
You can ensure proper prioritization if you separate this into a `tryObtainResource` and `obtainResource`. If you fail to obtain the resource immediately, return `OperationStatus.Waiting`, and when you successfully obtain the resource, set the status back to `OperationStatus.Ready`. You'll want to do a tiny bit of extra tracking so that when the operation gets requeued it picks up the already reserved resource, but that's straightforward.

Alternatively, you can skip the `tryObtainResource` and always have an operation initially return `OperationStatus.Waiting`, unless your synchronous cache of assigned resources says the operation already has one assigned, which is pretty much the same thing.